mlir/test/Dialect/Linalg/canonicalize.mlir

```diff
@@ -1232,3 +1232,21 @@ func.func @transpose_buffer(%input: memref<?xf32>,
 //  CHECK-SAME:            %[[VAL_1:.*]]: memref<?xf32>) {
 //       CHECK:     linalg.transpose ins(%[[VAL_0]] : memref<?xf32>)
 //  CHECK-SAME:       outs(%[[VAL_1]] : memref<?xf32>) permutation = [0]
+
+// -----
+
+// This test checks linalg op has a recursive memory effect. Otherwise
+// linalg.map without a user would be DCEd.
+func.func @recursive_effect(%arg0: memref<1xf32>, %arg1 : tensor<1xf32>) {
+  %c1 = arith.constant 1 : index
+  %init = arith.constant dense<0.0> : tensor<1xf32>
+  %mapped = linalg.map ins(%arg1:tensor<1xf32>) outs(%init :tensor<1xf32>)
+            (%in : f32) {
+              memref.store %in, %arg0[%c1] : memref<1xf32>
```

Comment:
Could we use tensors all the way here?

Reply:
We need an operation with memory side effects here, I'am afraid using tensors only cannt achieve this.

Reply:
Perhaps there are some usages that I'm unaware of. I'd be happy to hear if you could suggest any.

Reply:
Doesn't really make a difference, but a `vector.print %in : f32` should also work.

Reply:
Good point, done!~